engine/src/flutter/shell/platform/windows/flutter_host_window_controller.cc

```diff
@@ -0,0 +1,228 @@
+
+#include <dwmapi.h>
+#include <optional>
+#include <vector>
+
+#include "embedder.h"
+#include "flutter/shell/platform/common/windowing.h"
+#include "flutter/shell/platform/windows/flutter_windows_engine.h"
+#include "flutter/shell/platform/windows/flutter_windows_view_controller.h"
+#include "fml/logging.h"
+#include "shell/platform/windows/client_wrapper/include/flutter/flutter_view.h"
+#include "shell/platform/windows/flutter_host_window.h"
+#include "shell/platform/windows/flutter_windows_view.h"
+
+namespace flutter {
+
+struct WindowingInitRequest {
+  void (*on_message)(WindowsMessage*);
+};
+
+struct WindowCreationRequest {
+  double width;
+  double height;
+  double min_width;
+  double min_height;
+  double max_width;
+  double max_height;
+};
+}  // namespace flutter
+
+extern "C" {
+
+FLUTTER_EXPORT
+void flutter_windowing_initialize(
+    int64_t engine_id,
+    const flutter::WindowingInitRequest* request) {
+  flutter::FlutterWindowsEngine* engine =
+      flutter::FlutterWindowsEngine::GetEngineForId(engine_id);
+  engine->get_host_window_controller()->Initialize(request);
+}
+
+FLUTTER_EXPORT
+bool flutter_windowing_has_top_level_windows(int64_t engine_id) {
+  flutter::FlutterWindowsEngine* engine =
+      flutter::FlutterWindowsEngine::GetEngineForId(engine_id);
+  return engine->get_host_window_controller()->HasTopLevelWindows();
+}
+
+FLUTTER_EXPORT
+int64_t flutter_create_regular_window(
+    int64_t engine_id,
+    const flutter::WindowCreationRequest* request) {
+  flutter::FlutterWindowsEngine* engine =
+      flutter::FlutterWindowsEngine::GetEngineForId(engine_id);
+  return engine->get_host_window_controller()->CreateRegularWindow(request);
+}
+
+FLUTTER_EXPORT
+HWND flutter_get_window_handle(int64_t engine_id, FlutterViewId view_id) {
+  flutter::FlutterWindowsEngine* engine =
+      flutter::FlutterWindowsEngine::GetEngineForId(engine_id);
+  flutter::FlutterWindowsView* view = engine->view(view_id);
+  if (view == nullptr) {
+    return nullptr;
+  } else {
+    return GetAncestor(view->GetWindowHandle(), GA_ROOT);
+  }
+}
+
+struct Size {
+  double width;
+  double height;
+};
+
+FLUTTER_EXPORT
+void flutter_get_window_size(HWND hwnd, Size* size) {
+  RECT rect;
+  GetClientRect(hwnd, &rect);
+  double const dpr = FlutterDesktopGetDpiForHWND(hwnd) /
+                     static_cast<double>(USER_DEFAULT_SCREEN_DPI);
+  double const width = rect.right / dpr;
+  double const height = rect.bottom / dpr;
+  size->width = width;
+  size->height = height;
+}
+
+FLUTTER_EXPORT
+int64_t flutter_get_window_state(HWND hwnd) {
+  if (IsIconic(hwnd)) {
+    return static_cast<int64_t>(flutter::WindowState::kMinimized);
+  } else if (IsZoomed(hwnd)) {
+    return static_cast<int64_t>(flutter::WindowState::kMaximized);
+  } else {
+    return static_cast<int64_t>(flutter::WindowState::kRestored);
+  }
+}
+
+FLUTTER_EXPORT
+void flutter_set_window_state(HWND hwnd, int64_t state) {
```

Comment:
Is there a reason to implement this in C++ instead of Dart? Did you consider FFI'ing straight to win32? Something like this: https://github.com/halildurmus/win32/blob/27496b847b9ce831db798203238e4aa847e87d54/packages/win32/lib/src/win32/user32.g.dart#L9200-L9214

Reply:
For trivial things I already call win32 API directly, for example `_setWindowTitle` calls `SetWindowTextW`, `_destroyWindow` calls `DestroyWindow`. Here i went this way because it seemed simpler (needs less declarations and constants), in reality there is nothing preventing us from calling win32 api directly.

Reply:
That's a completely reasonable answer, but I suspect that more lines of code is a good trade-off for moving more of the logic to Dart.

Customers know how to debug Dart code. However, they might be significantly less comfortable debugging C++ code. The more logic is in the Dart layer, the more accessible that logic is to our customers.

Of course, this is a balancing act. If there are areas where Dart FFI to win32 is significantly less readable/maintainable, we should stick to C++.